### ISSUE TITLE
Use Dynamic Fee for More than Just `attach` (?)

### ISSUE THREAD

**Issue by adamkrellenstein** [CounterpartyXCP/counterparty-core#2893]
When Counterparty was first created, the XCP fees for various transactions were hard-coded to low, constant values (e.g. 0.5 XCP for issuing a named asset). This system was chosen for its simplicity, but it creates *significant friction* when onboarding users to the Counterparty ecosystem (esp. since it is so hard to acquire XCP) and yet the fees are so low that they have little economic significance for market participants.

A proper XCP fee system should be proportional to network traffic and transaction complexity while reducing unnecessary barriers to entry into the network. When transaction volumes are low, the fee should be zero; when they're high, they should go up quickly. This should simultaneously reduce the barriers to entry into Counterparty and increase long-term demand for XCP.

We've already implemented such a system for `attach` transactions—it's described here: https://docs.counterparty.io/docs/advanced/specifications/gas-system/ This issue is a proposal to use the above gas system not just for `attach` transactions but also for named asset issuances, etc.

**Comment by droplister** [comment]
Has the gas system for `attach` risen to a level that has trigged the gas system? I would like to see it work in action there first. But I like the idea of transitioning XCP from a sort of registration fee for assets to a gas for transactions.

Is the long term goal to put a dynamic gas fee on all transactions? I think that would be preferrable to having a mixture of flat fees (asset issuance), "per row" fees (dividends, sweeps), and dynamic fees (attach).

In that case, you could have a single gas fee endpoint and you could introduce to the gas system a `complexity` factor that's like 1 for basic actions like sends, attach, and 10 or something for dividends and sweeps (if you wanted to make dividends and sweeps cost calculations simpler while still being higher than basic actions.) There's merit to continuing to use the # of holders or # of assets in those calculations still, just an idea.

**Comment by droplister** [comment]
Sounds based.

**Comment by zerog975** [comment]
I disagree. This is creating a punitive system where counterparty users would actually see an increase fees as usage/demand increases. 
This is on top of the cost of the BTC fees users always have to pay.
These incentives would harm and limit adoption. 
Don't forget, XCP isn't the only game in town. If users feel fees are excessive or just unfair, they can instead use many of the other options, such as Ordinals, ETH (gas is 1 right now), SOL, etc.
I have yet to hear users asking for higher fees.


A system where users have to have a secondary app or website to check the current virtual "fee" market for their TXs is a worse user experience, or friction, than as it is. There is no need to create a scaling fee system, as the BTC TX capacity and fees prevent excessive spam.
The current system is deflationary over time, putting excessive focus on XCP burn/demand isn't appropriate.
I think increasing demand for XCP shouldn't be a goal at this point. 
If that is a goal, demand for XCP will increase as more users start using counterparty. More users, more demand.  Why is that model not sufficient?

If counterparty gets too cute with fees and tokenomics, it risks damaging its credibility, which isn't easily recovered.

**Comment by davestaxcp** [comment]
yeah confused here how the thesis of this starts that the current system "creates significant friction when onboarding users to the Counterparty ecosystem" but then the solution is adding more xcp fees?

if users want to create numeric assets, there is no xcp fee.... thesis of that has always been you dont need xcp to use the protocol

also as ZeroG mentioned, our friction is simply BTC fees for simple sends, creating XCP DEX orders, MPMA etc.... how is it still a bitcoin meta-protocol as stated in Adam K's recent whitepaper with xcp fees added to everything?

are XCP holders the only people benefiting from fees here? because it definetely wont be new users

furthermore I assume the reasoning "well the fees will be 0 if no activity exists" is somewhat shallow because once there is even a FRACTION of fees to use counterparty, there will be incredible and instant friction for any new user - especially because XCP is difficult to get at this moment (only dispensers, the crappy DEX-TRADE cex and limited PSBT/Atomic Swaps markets at this time)

dont really understand nor agree with the reasoning that "This should simultaneously reduce the barriers to entry into Counterparty and increase long-term demand for XCP"

**Comment by adamkrellenstein** [comment]
Not sure why this proposal is understood to "increase XCP fees". Fees would only be higher than current levels after high levels of adoption _have already been reached_. Fees will be exactly zero until that point. If the concern is that these fees could _limit_ adoption, then the parameters simply need to be set accordingly. A priori a dynamic fee system is strictly better than a fixed one. And a priori BTC fees cannot be relied upon exclusively insofar as the Counterparty protocol is a significant extension to Bitcoin itself.

**Comment by zerog975** [comment]
quoting "when they're high, they should go up quickly. This should simultaneously reduce the barriers to entry into Counterparty and increase long-term demand for XCP"

Why not scale from 0 to the current fixed fee then?
It appears that the intention is to increase the demand for XCP, as you stated.

**Comment by davestaxcp** [comment]
we have fees on features like

- creating an xcp asset
- doing a sweep
- wrapped a token into a UTXO

which frankly, im fine with - they are great features and things like sweep using dynamic xcp fees make complete sense due to how much data the transaction needs (see https://github.com/CounterpartyXCP/counterparty-core/issues/1220#issuecomment-1747037727)


are suggesting you want dynamic fees on features like

- sending tokens?
- creating DEX orders?
- creating Dispensers?


lets say https://github.com/CounterpartyXCP/counterparty-core/issues/2749 goes through and we have hundreds of thousands of users etc..... and now everyone needs to use tiny amount of xcp to do all functions (similar to Arweave)

![Image](https://github.com/user-attachments/assets/2d68f0ce-cef1-459a-b5df-cdaa621fec45)

how many xcp transactions are we looking at for the beginning of these fees to kick in as stated in the graph above (in the 9 to 11 transactions per block range)?

also im confused on reading the docs link you mentioned "function of the number of transactions using this function" so this fee system is separate for each function used?

for instance if DEX orders somehow just were blowing up and many users were making them, at what point does it need an xcp fee to use? arent these users making these dex orders going to get used to having 0 fees and then randomly they have to pay all of a sudden? at what point does that happen in the graph included above?

**Comment by davestaxcp** [comment]
maybe each function itself needs a separate discussion on xcp fees?

since the protocol posts recently specically eyed to remove fees for subassets too (which agree with!)

**Comment by adamkrellenstein** [comment]
> Why not scale from 0 to the current fixed fee then?

Because scaling to a fixed value would handicap the economic incentives associated with having a fee system in the first place. Imagine if Bitcoin transaction fees were capped at a fixed value... Rather, the constant associated with the transaction volume beyond which the fees start to increase could be arbitrarily high, which would lead to ~the same outcome. (Instead of 40 tx/block in the above graph, it could be 4000 tx / block e.g.)

----------

Nowhere in the proposal is it suggested that fees should be added to transactions that do not already have them. I think that would be a bad idea in general.

> also im confused on reading the docs link you mentioned "function of the number of transactions using this function" so this fee system is separate for each function used?

Yes, the constants should be different for each transaction type, and the transaction volume calculated independently for each type. (Lots of name asset issuances should not increase the fees for e.g. attaches.) They can all be discussed and determined independently.

**Comment by zerog975** [comment]
> > Why not scale from 0 to the current fixed fee then?
> 
> Because scaling to a fixed value would handicap the economic incentives associated with having a fee system in the first place. Imagine if Bitcoin transaction fees were capped at a fixed value... Rather, the constant associated with the transaction volume beyond which the fees start to increase could be arbitrarily high, which would lead to ~the same outcome. (Instead of 40 tx/block in the above graph, it could be 4000 tx / block e.g.)
> 
> Nowhere in the proposal is it suggested that fees should be added to transactions that do not already have them. I think that would be a bad idea in general.
> 
> > also im confused on reading the docs link you mentioned "function of the number of transactions using this function" so this fee system is separate for each function used?
> 
> Yes, the constants should be different for each transaction type, and the transaction volume calculated independently for each type. (Lots of name asset issuances should not increase the fees for e.g. attaches.) They can all be discussed and determined independently.

What is the point of this proposal?  
It was stated before that it was to increase demand for XCP.
TXs are already capped by the low capacity of BTC itself.
If the point is to not increase overall fees, then capping it at the current fee works, even though it would create more complexity.
But if the goal is to increase fees, then that is why you would create a more complicated system.

**Comment by davestaxcp** [comment]
i can understand the dynamic change for functions where we are already used to fees (as i mentioned before)

but yes adding fees to existing functions definetely should be discussed and determined independently for each function

as I think some functions having fees would make sense while other i think have little to no merit or user support for fees at all

especially because the original thesis from my understanding with early XCP fees for Named Assets was for a "non spam fee"... so people wouldnt create names and sit on them similar to domain name registrations or early namecoin registrations?

**Comment by honma-munehisa** [comment]
I support the proposed move to a dynamic XCP fee system, and I hope to see it implemented in the near future.

**Comment by jdogresorg** [comment]
I feel that the features which have fixed prices such as asset registration should stay at fixed prices (0.50 for assets and 0.25 for subassets). The justification for lowering the fee amount in order to keep the cost of asset registrations low if/when the price of XCP rises, makes no sense to me considering there is a proposal to 100x or 1000x the XCP supply, which will drastically inflate the supply and bring cost of XCP down considerably already. As the largest asset name holder on the network, I do not support this change.

NACK - I oppose this dynamic fee change on asset and sub-asset registrations.

I see 3+ people raising concerns about this switch and not much support for it. I hope that the developers listen to the community on this issue instead of continue to force contentious changes.

**Comment by vectorconfetti** [comment]
Would it be the case that asset issuance fees will typically be 0 if there isn’t a lot of that type of transaction in recent blocks, similar to how UTXO attaches work?

**Comment by vectorconfetti** [comment]
Just to add to my thumbs up - I think this is a good idea and is very much in line with how fees are done in the rest of the crypto fee ecosystem (including BTC). Fixed XCP fees are the odd man out here.

**Comment by adamkrellenstein** [comment]
Yes @vectorconfetti the goal is to make the fee exactly zero, rather than "low". The latter is both economically useless and a needless barrier to entry into the ecosystem.

Note: There is no proposal to increase the supply of XCP of course. The proposal is to change the default unit—to move the decimal place—which would of course change the nominal fee quantity as well.

**Comment by pappyg45** [comment]
I oppose this dynamic fee change on asset and sub-asset registrations

**Comment by vectorconfetti** [comment]
> I support dynamic fees being implemented everywhere else where we charge fees currently, except in the case of asset issuances.

Can you elaborate on why asset issuances should have a different fee structure than every other transaction with fees on the network?

**Comment by pappyg45** [comment]
> > I support dynamic fees being implemented everywhere else where we charge fees currently, except in the case of asset issuances.
> 
> Can you elaborate on why asset issuances should have a different fee structure than every other transaction with fees on the network?

there is already free numerics so it makes no sense, vanity ownership of assets should remain consistent as the historical costs associated with this anti-spam mechanism should remain.

**Comment by pappyg45** [comment]
will existing users who issued named assets in the past be reimbursed with XCP for the newer fee structure?  asset issuance like is similar to the block rewards in bitcoin. as time goes on it get harder to mine new blocks with less reward so in keeping like bitcoin asset issuance over time could have a halving schedule which I would support but as it stands this is nonsense. Variable fees on all else is ok.  again we already have free numerics

**Comment by vectorconfetti** [comment]
> as time goes on it get harder to mine new blocks with less reward so in keeping like bitcoin asset issuance over time could have a halving schedule which I would support but as it stands this is nonsense.

Block subsidy rewards (new supply) are different than fees. The halvening controls inflation. They are provided to people running expensive nodes and paying for power, not for users of the network issuing transactions.

Bitcoin fees are charged to users, and as you know, are dynamic based on usage. Exactly as proposed here.

**Comment by pappyg45** [comment]
> > as time goes on it get harder to mine new blocks with less reward so in keeping like bitcoin asset issuance over time could have a halving schedule which I would support but as it stands this is nonsense.
> 
> Block subsidy rewards (new supply) are different than fees. The halvening controls inflation. They are provided to people running expensive nodes and paying for power, not for users of the network issuing transactions.
> 
> Bitcoin fees are charged to users, and as you know, are dynamic based on usage. Exactly as proposed here.

you are ignoring physics. you are conflating fees with the foundation of this protocol...

Burned bitcoin = xcp = named asset

bitcoin subsidy trend toward zero of fixed supply

asset issuance (burn) of fixed supply trend towards zero
